File: impl/swift/Sources/Naalp/Records.swift
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
//
// N-AALP body builders for the Swift SDK — the deterministic-CBOR bodies of the spine records:
// approval + consume ledger (C6, §7), audit receipt (C7, §8), delivery update (C8, §9), stream
// open/commit/checkpoint + rolling commitment (C9, §10), foreign carriage (C12, §13), and the
// transport confidentiality boundary (C11, §12). Each body is exactly what the Go, Rust and
// Python reference implementations encode, so the bytes are byte-identical.

import Crypto
import Foundation

public enum Records {

    // --- C6 approval + consume ledger (§7) ---

    public static func approvalBody(_ approves: [UInt8], _ approver: String, _ grant: UInt64,
                                    _ nonce: [UInt8], _ notAfter: UInt64) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .b(approves)),
            (.u(2), .t(approver)),
            (.u(3), .u(grant)),
            (.u(4), .b(nonce)),
            (.u(5), .u(notAfter)),
        ]))
    }

    public static func approvalId(_ approves: [UInt8], _ approver: String, _ grant: UInt64,
                                  _ nonce: [UInt8], _ notAfter: UInt64) throws -> [UInt8] {
        return Cbor.contentId(try approvalBody(approves, approver, grant, nonce, notAfter))
    }

    public static func ledgerEntry(_ seq: UInt64, _ prev: [UInt8], _ approvalId: [UInt8], _ by: String) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .u(seq)),
            (.u(2), .b(prev)),
            (.u(3), .b(approvalId)),
            (.u(4), .t(by)),
        ]))
    }

    // --- C7 audit receipt (§8) ---

    public static func receiptBody(_ prev: [UInt8], _ obj: [UInt8], _ seq: UInt64, _ at: UInt64) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .b(prev)),
            (.u(2), .b(obj)),
            (.u(3), .u(seq)),
            (.u(4), .u(at)),
        ]))
    }

    public static func receiptHead(_ body: [UInt8]) -> [UInt8] {
        return Array(SHA384.hash(data: Data(body)))
    }

    // --- C8 delivery (§9) ---

    public static func deliveryUpdate(_ obj: [UInt8], _ stage: UInt64, _ at: UInt64) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .b(obj)),
            (.u(2), .u(stage)),
            (.u(3), .u(at)),
        ]))
    }

    // --- C9 streaming (§10) ---

    /// Rolling SHA-384 over the chunk data in absolute-offset order (R-10.2).
    public static func streamDigest(_ chunks: [(offset: UInt64, data: [UInt8])]) -> [UInt8] {
        let ordered = chunks.sorted { $0.offset < $1.offset }
        var hasher = SHA384()
        for c in ordered {
            hasher.update(data: Data(c.data))
        }
        return Array(hasher.finalize())
    }

    public static func streamOpenBody(_ streamId: [UInt8], _ effect: UInt64, _ approval: [UInt8], _ substream: UInt64) throws -> [UInt8] {
        var pairs: [(CborValue, CborValue)] = [
            (.u(1), .b(streamId)),
            (.u(2), .u(effect)),
            (.u(4), .u(substream)),
        ]
        if !approval.isEmpty {  // field 3 present only when an approval binding exists
            pairs.append((.u(3), .b(approval)))
        }
        return try Cbor.encode(.m(pairs))
    }

    public static func streamCommitBody(_ streamId: [UInt8], _ digest: [UInt8]) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .b(streamId)),
            (.u(2), .b(digest)),
        ]))
    }

    public static func streamCheckpointBody(_ streamId: [UInt8], _ throughOffset: UInt64, _ digestSoFar: [UInt8]) throws -> [UInt8] {
        return try Cbor.encode(.m([
            (.u(1), .b(streamId)),
            (.u(2), .u(throughOffset)),
            (.u(3), .b(digestSoFar)),
        ]))
    }

    // --- C12 foreign carriage (§13) ---

    public static let CLASS_JSONRPC: UInt64 = 0
    public static let CLASS_HTTP: UInt64 = 1
    public static let CLASS_MSG: UInt64 = 2
    public static let CLASS_STREAM: UInt64 = 3
    public static let CLASS_DOC: UInt64 = 4
    public static let CLASS_OPAQUE: UInt64 = 5

    public static func carriageBody(_ protocolId: UInt64, _ cls: UInt64, _ contentType: UInt64,
                                    _ correlation: [UInt8], _ method: String, _ foreign: [UInt8]) throws -> [UInt8] {
        if cls > CLASS_OPAQUE {
            throw NaalpError("MappingError", "carriage class \(cls) is not defined")
        }
        return try Cbor.encode(.m([
            (.u(1), .u(protocolId)),
            (.u(2), .u(cls)),
            (.u(3), .u(contentType)),
            (.u(4), .b(correlation)),
            (.u(5), .t(method)),
            (.u(6), .b(foreign)),
        ]))
    }

    // --- C11 transport confidentiality boundary (§12) ---

    // (confidential, peer_authenticated)
    static let transports: [String: (Bool, Bool)] = [
        "npamp": (true, true),
        "quic": (true, true),
        "websocket+wss": (true, false),
        "websocket+ws": (false, false),
        "https": (true, false),
        "http": (false, false),
    ]

    /// Apply the §12.3 confidentiality boundary + §12.4 peer-auth rule; returns the result label.
    public static func transportEmit(_ name: String, _ sensitive: Bool, _ requirePeerAuth: Bool) throws -> String {
        guard let t = transports[name] else {
            throw NaalpError("UnknownTransport", "unknown transport \(name)")
        }
        let (confidential, peerAuthenticated) = t
        if sensitive && !confidential {
            return "ConfidentialTransportRequired"
        }
        if requirePeerAuth && !peerAuthenticated {
            return "PeerUnauthenticated"
        }
        return "ok"
    }
}

File: impl/ruby/lib/naalp/records.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# N-AALP body builders for the Ruby SDK — the deterministic-CBOR bodies of the spine records:
# approval + consume ledger (C6, §7), audit receipt (C7, §8), delivery update (C8, §9), stream
# open/commit/checkpoint + rolling commitment (C9, §10), foreign carriage (C12, §13), and the
# transport confidentiality boundary (C11, §12). Each body is exactly what the Go, Rust and
# Python reference implementations encode, so the bytes are byte-identical.
require 'openssl'
require_relative 'cbor'

module Naalp
  module Records
    U = Naalp::CBOR::U
    B = Naalp::CBOR::B
    T = Naalp::CBOR::T
    M = Naalp::CBOR::M

    # --- C6 approval + consume ledger (§7) ---

    def self.approval_body(approves, approver, grant, nonce, not_after)
      Naalp::CBOR.encode(M.new([
        [U.new(1), B.new(approves)], [U.new(2), T.new(approver)], [U.new(3), U.new(grant)],
        [U.new(4), B.new(nonce)], [U.new(5), U.new(not_after)],
      ]))
    end

    def self.approval_id(approves, approver, grant, nonce, not_after)
      Naalp::CBOR.content_id(approval_body(approves, approver, grant, nonce, not_after))
    end

    def self.ledger_entry(seq, prev, approval_id_bytes, by)
      Naalp::CBOR.encode(M.new([
        [U.new(1), U.new(seq)], [U.new(2), B.new(prev)], [U.new(3), B.new(approval_id_bytes)], [U.new(4), T.new(by)],
      ]))
    end

    # --- C7 audit receipt (§8) ---

    def self.receipt_body(prev, obj, seq, at)
      Naalp::CBOR.encode(M.new([
        [U.new(1), B.new(prev)], [U.new(2), B.new(obj)], [U.new(3), U.new(seq)], [U.new(4), U.new(at)],
      ]))
    end

    def self.receipt_head(body)
      OpenSSL::Digest::SHA384.digest(body)
    end

    # --- C8 delivery (§9) ---

    def self.delivery_update(obj, stage, at)
      Naalp::CBOR.encode(M.new([[U.new(1), B.new(obj)], [U.new(2), U.new(stage)], [U.new(3), U.new(at)]]))
    end

    # --- C9 streaming (§10) ---

    # Rolling SHA-384 over the chunk data in absolute-offset order (R-10.2).
    def self.stream_digest(chunks)
      h = OpenSSL::Digest::SHA384.new
      chunks.sort_by { |offset, _data| offset }.each { |_offset, data| h.update(data) }
      h.digest
    end

    def self.stream_open_body(stream_id, effect, approval, substream)
      pairs = [[U.new(1), B.new(stream_id)], [U.new(2), U.new(effect)], [U.new(4), U.new(substream)]]
      pairs << [U.new(3), B.new(approval)] if approval && approval.bytesize > 0
      Naalp::CBOR.encode(M.new(pairs))
    end

    def self.stream_commit_body(stream_id, digest)
      Naalp::CBOR.encode(M.new([[U.new(1), B.new(stream_id)], [U.new(2), B.new(digest)]]))
    end

    def self.stream_checkpoint_body(stream_id, through_offset, digest_so_far)
      Naalp::CBOR.encode(M.new([
        [U.new(1), B.new(stream_id)], [U.new(2), U.new(through_offset)], [U.new(3), B.new(digest_so_far)],
      ]))
    end

    # --- C12 foreign carriage (§13) ---

    CLASS_JSONRPC = 0
    CLASS_HTTP = 1
    CLASS_MSG = 2
    CLASS_STREAM = 3
    CLASS_DOC = 4
    CLASS_OPAQUE = 5

    class MappingError < StandardError
      def kind; "MappingError"; end
    end

    def self.carriage_body(protocol_id, cls, content_type, correlation, method_name, foreign)
      raise MappingError, "carriage class #{cls} is not defined" if cls > CLASS_OPAQUE
      Naalp::CBOR.encode(M.new([
        [U.new(1), U.new(protocol_id)], [U.new(2), U.new(cls)], [U.new(3), U.new(content_type)],
        [U.new(4), B.new(correlation)], [U.new(5), T.new(method_name)], [U.new(6), B.new(foreign)],
      ]))
    end

    # --- C11 transport confidentiality boundary (§12) ---
    # value: [confidential, peer_authenticated]
    TRANSPORTS = {
      "npamp" => [true, true],
      "quic" => [true, true],
      "websocket+wss" => [true, false],
      "websocket+ws" => [false, false],
      "https" => [true, false],
      "http" => [false, false],
    }.freeze

    class UnknownTransport < StandardError
      def kind; "UnknownTransport"; end
    end

    # Apply the §12.3 confidentiality boundary + §12.4 peer-auth rule; returns the result label.
    def self.transport_emit(name, sensitive, require_peer_auth)
      t = TRANSPORTS[name]
      raise UnknownTransport, "unknown transport #{name.inspect}" if t.nil?
      confidential, peer_authenticated = t
      return "ConfidentialTransportRequired" if sensitive && !confidential
      return "PeerUnauthenticated" if require_peer_auth && !peer_authenticated
      "ok"
    end
  end
end
